File: tools/generate-text-codec-tables.py
#!/usr/bin/env python3
"""Pack the revision-pinned WebKit decoder indexes used by TextCodec.

The inputs are the unmodified EncodingTables.cpp, TextCodecSingleByte.cpp, and
TextCodecCJK.cpp files from home-lang/home revision
7ed99c02e50034f869d0db6d487115bb44332fe4.
Normal builds consume the checked-in binary and never run this generator.
"""

from __future__ import annotations

import argparse
import hashlib
import re
import struct
from pathlib import Path

ENCODING_TABLES_SHA256 = "ec3ec297dd8b52a64ac3065206d26ab186b9f59349790d7e6bfcda297051ad60"
SINGLE_BYTE_SHA256 = "6abca8d12653ab4dde31ad3eae0e089b392b793d7a9bc04c68081fb6536b8552"
CJK_SHA256 = "d178f1c382a2c590f7d812ff317d9b20e5085358604eeb4eb4b40f3697704508"
MAGIC = b"ZJTC0001"

SINGLE_BYTE_TABLES = (
    "iso88593",
    "iso88596",
    "iso88597",
    "iso88598",
    "windows874",
    "windows1253",
    "windows1255",
    "windows1257",
    "koi8u",
    "ibm866",
)


def checked_source(path: Path, expected_sha256: str) -> str:
    data = path.read_bytes()
    actual = hashlib.sha256(data).hexdigest()
    if actual != expected_sha256:
        raise SystemExit(f"{path}: expected SHA-256 {expected_sha256}, got {actual}")
    return data.decode()


def initializer(source: str, marker: str) -> str:
    start = source.index(marker)
    start = source.index("{", start) + 1
    end = source.index("};", start)
    return source[start:end]


def pairs(source: str, marker: str, expected_count: int) -> list[tuple[int, int]]:
    result = [
        (int(key, 10), int(value, 16))
        for key, value in re.findall(r"\{\s*(\d+)\s*,\s*(0x[0-9a-fA-F]+)\s*\}", initializer(source, marker))
    ]
    if len(result) != expected_count:
        raise SystemExit(f"{marker}: expected {expected_count} entries, got {len(result)}")
    return result


def hex_values(source: str, marker: str, expected_count: int) -> list[int]:
    result = [int(value, 16) for value in re.findall(r"0x[0-9a-fA-F]+", initializer(source, marker))]
    if len(result) != expected_count:
        raise SystemExit(f"{marker}: expected {expected_count} entries, got {len(result)}")
    return result


def dense(entries: list[tuple[int, int]], count: int) -> list[int]:
    result = [0] * count
    for pointer, code_point in entries:
        if pointer >= count or result[pointer] != 0:
            raise SystemExit(f"invalid or duplicate pointer {pointer}")
        result[pointer] = code_point
    return result


def write_u16(output: bytearray, values: list[int]) -> None:
    for value in values:
        output.extend(struct.pack("<H", value))


def write_u32(output: bytearray, values: list[int]) -> None:
    for value in values:
        output.extend(struct.pack("<I", value))


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("encoding_tables", type=Path)
    parser.add_argument("single_byte", type=Path)
    parser.add_argument("cjk", type=Path)
    parser.add_argument("output", type=Path)
    args = parser.parse_args()

    encoding = checked_source(args.encoding_tables, ENCODING_TABLES_SHA256)
    single = checked_source(args.single_byte, SINGLE_BYTE_SHA256)
    cjk = checked_source(args.cjk, CJK_SHA256)
    output = bytearray(MAGIC)

    for name in SINGLE_BYTE_TABLES:
        write_u16(output, hex_values(single, f"SingleByteDecodeTable {name}", 128))

    write_u16(output, dense(pairs(encoding, "jis0208Data", 7724), 11104))
    write_u16(output, dense(pairs(encoding, "jis0212Data", 6067), 7211))
    write_u32(output, dense(pairs(encoding, "big5Data", 18590), 19782))
    write_u16(output, dense(pairs(encoding, "eucKRData", 17048), 23750))
    write_u16(output, hex_values(encoding, "gb18030Data", 23940))

    ranges = pairs(cjk, "gb18030Ranges", 207)
    write_u32(output, [value for pair in ranges for value in pair])

    args.output.parent.mkdir(parents=True, exist_ok=True)
    args.output.write_bytes(output)
    print(f"wrote {len(output)} bytes to {args.output}")


if __name__ == "__main__":
    main()

File: tools/generate-text-codec-tables.ts
/** Pack revision-pinned WebKit TextCodec decoder indexes. */
import { readText, run, writeHex } from "./lib/home";

const expected = [
  "ec3ec297dd8b52a64ac3065206d26ab186b9f59349790d7e6bfcda297051ad60",
  "6abca8d12653ab4dde31ad3eae0e089b392b793d7a9bc04c68081fb6536b8552",
  "d178f1c382a2c590f7d812ff317d9b20e5085358604eeb4eb4b40f3697704508",
];
const names = ["iso88593", "iso88596", "iso88597", "iso88598", "windows874", "windows1253", "windows1255", "windows1257", "koi8u", "ibm866"];
const args = process.argv.slice(2);
if (args.length !== 4) throw new Error("usage: generate-text-codec-tables.ts ENCODING_TABLES SINGLE_BYTE CJK OUTPUT");
function checkedSource(path: string, digest: string): string {
  const result = run(["shasum", "-a", "256", path]);
  if (result.exitCode !== 0) throw new Error(result.stderr || "cannot checksum " + path);
  const actual = result.stdout.trim().split(/\s+/)[0];
  if (actual !== digest) throw new Error(`${path}: expected SHA-256 ${digest}, got ${actual}`);
  return readText(path);
}
const encoding = checkedSource(args[0], expected[0]);
const single = checkedSource(args[1], expected[1]);
const cjk = checkedSource(args[2], expected[2]);
function initializer(source: string, marker: string): string {
  let start = source.indexOf(marker);
  if (start < 0) throw new Error("missing initializer: " + marker);
  start = source.indexOf("{", start) + 1;
  const end = source.indexOf("};", start);
  if (end < 0) throw new Error("unterminated initializer: " + marker);
  return source.slice(start, end);
}
function pairs(source: string, marker: string, count: number): number[][] {
  const rows: number[][] = [], regex = /\{\s*(\d+)\s*,\s*(0x[0-9a-fA-F]+)\s*\}/g;
  const body = initializer(source, marker);
  let match: RegExpExecArray | null;
  while ((match = regex.exec(body)) !== null) rows.push([Number(match[1]), Number.parseInt(match[2], 16)]);
  if (rows.length !== count) throw new Error(`${marker}: expected ${count} entries, got ${rows.length}`);
  return rows;
}
function hexValues(source: string, marker: string, count: number): number[] {
  const values: number[] = [], regex = /0x[0-9a-fA-F]+/g;
  const body = initializer(source, marker);
  let match: RegExpExecArray | null;
  while ((match = regex.exec(body)) !== null) values.push(Number.parseInt(match[0], 16));
  if (values.length !== count) throw new Error(`${marker}: expected ${count} entries, got ${values.length}`);
  return values;
}
function dense(entries: number[][], count: number): number[] {
  const values = new Array(count).fill(0);
  for (const [pointer, point] of entries) {
    if (pointer >= count || values[pointer] !== 0) throw new Error("invalid or duplicate pointer " + pointer);
    values[pointer] = point;
  }
  return values;
}
const bytes: number[] = [0x5a, 0x4a, 0x54, 0x43, 0x30, 0x30, 0x30, 0x31];
const u16 = (values: number[]) => { for (const value of values) bytes.push(value & 255, (value >>> 8) & 255); };
const u32 = (values: number[]) => { for (const value of values) bytes.push(value & 255, (value >>> 8) & 255, (value >>> 16) & 255, (value >>> 24) & 255); };
for (const name of names) u16(hexValues(single, "SingleByteDecodeTable " + name, 128));
u16(dense(pairs(encoding, "jis0208Data", 7724), 11104));
u16(dense(pairs(encoding, "jis0212Data", 6067), 7211));
u32(dense(pairs(encoding, "big5Data", 18590), 19782));
u16(dense(pairs(encoding, "eucKRData", 17048), 23750));
u16(hexValues(encoding, "gb18030Data", 23940));
u32(pairs(cjk, "gb18030Ranges", 207).reduce((all: number[], pair: number[]) => all.concat(pair), []));
const slash = args[3].lastIndexOf("/");
if (slash > 0) {
  const made = run(["mkdir", "-p", args[3].slice(0, slash)]);
  if (made.exitCode !== 0) throw new Error(made.stderr || "cannot create output directory");
}
writeHex(args[3], bytes.map(byte => byte.toString(16).padStart(2, "0")).join(""));
console.log(`wrote ${bytes.length} bytes to ${args[3]}`);
